torch/custom_class.h

```diff
@@ -1,273 +1,2 @@
-    auto arg_type = setstate_schema.arguments().at(1).type();
-    TORCH_CHECK(
-        (*arg_type == *ser_type),
-        "__setstate__'s argument should be the same type as the "
-        "return value of __getstate__. Got ",
-        arg_type->python_str(),
-        " but expected ",
-        ser_type->python_str());
-
-    return *this;
-  }
-
- private:
-  template <typename Func>
-  void defineMethod(std::string name, Func func) {
-    auto qualMethodName = qualClassName + "." + name;
-    auto schema = c10::inferFunctionSchemaSingleReturn<Func>(std::move(name), "");
-
-    auto wrapped_func = [func = std::move(func)](jit::Stack& stack) mutable -> void {
-      // TODO: we need to figure out how to profile calls to custom functions
-      // like this! Currently can't do it because the profiler stuff is in
-      // libtorch and not ATen
-      using RetType =
-          typename c10::guts::infer_function_traits_t<Func>::return_type;
-      detail::BoxedProxy<RetType, Func>()(stack, func);
-    };
-    auto method = std::make_unique<jit::BuiltinOpFunction>(
-        qualMethodName, std::move(schema), std::move(wrapped_func));
-
-    // Register the method here to keep the Method alive.
-    // ClassTypes do not hold ownership of their methods (normally it
-    // those are held by the CompilationUnit), so we need a proxy for
-    // that behavior here.
-    classTypePtr->addMethod(method.get());
-    registerCustomClassMethod(std::move(method));
-  }
-
-  std::string qualClassName;
-  at::ClassTypePtr classTypePtr;
-};
-
-/// make_custom_class() is a convenient way to create an instance of a registered
-/// custom class and wrap it in an IValue, for example when you want to pass the
-/// object to TorchScript. Its syntax is equivalent to APIs like `std::make_shared<>`
-/// or `c10::make_intrusive<>`.
-///
-/// For example, if you have a custom C++ class that can be constructed from an `int`
-/// and `std::string`, you might use this API like so:
-///
-///     IValue custom_class_iv = torch::make_custom_class<MyClass>(3, "foobarbaz");
-template <typename CurClass, typename... CtorArgs>
-c10::IValue make_custom_class(CtorArgs&&... args) {
-  if (!c10::isCustomClassRegistered<c10::intrusive_ptr<CurClass>>()) {
-    throw c10::Error(
-        "Trying to instantiate a class that isn't a registered custom class.",
-        "");
-  }
-  auto userClassInstance = c10::make_intrusive<CurClass>(std::forward<CtorArgs>(args)...);
-  return c10::IValue(std::move(userClassInstance));
-}
-
-// jit namespace for backward-compatibility
-// We previously defined everything in torch::jit but moved it out to
-// better reflect that these features are not limited only to TorchScript
-namespace jit {
-
-using ::torch::getCustomClass;
-using ::torch::isCustomClass;
-using ::torch::init;
-using ::torch::class_;
-
-} // namespace jit
-
-} // namespace torch
+#include <ATen/core/op_registration/custom_class.h>
```

Comment:
Do we really need to move all this? `custom_class.h` was meant to be easily accessible and readable for libtorch users and I don't like the indirection this is introducing

Reply:
I'm not sure. The operant question is whether or not I can include `torch/` base directory from `ATen/core`. I can *probably* make it work if I push on the build system enough, but my default assumption is it will require a bit of finagling. (Maybe less finagling since custom_class.h already works on mobile, though?)

Reply:
Should we ask if the same accessibility concerns apply to the new API as to `custom_class.h`? Maybe this is a symptom of the new stuff being too buried.